fe/fe-core/src/main/java/com/starrocks/sql/optimizer/SPMOptimizer.java

```diff
@@ -0,0 +1,319 @@
+import com.starrocks.sql.optimizer.rule.transformation.ConvertToEqualForNullRule;
+import com.starrocks.sql.optimizer.rule.transformation.EliminateAggRule;
+import com.starrocks.sql.optimizer.rule.transformation.EliminateConstantCTERule;
+import com.starrocks.sql.optimizer.rule.transformation.ForceCTEReuseRule;
+import com.starrocks.sql.optimizer.rule.transformation.JoinLeftAsscomRule;
+import com.starrocks.sql.optimizer.rule.transformation.MergeProjectWithChildRule;
+import com.starrocks.sql.optimizer.rule.transformation.MergeTwoAggRule;
+import com.starrocks.sql.optimizer.rule.transformation.MergeTwoProjectRule;
+import com.starrocks.sql.optimizer.rule.transformation.PruneEmptyWindowRule;
+import com.starrocks.sql.optimizer.rule.transformation.PushDownJoinOnExpressionToChildProject;
+import com.starrocks.sql.optimizer.rule.transformation.PushDownProjectLimitRule;
+import com.starrocks.sql.optimizer.rule.transformation.PushDownTopNBelowOuterJoinRule;
+import com.starrocks.sql.optimizer.rule.transformation.PushDownTopNBelowUnionRule;
+import com.starrocks.sql.optimizer.rule.transformation.PushLimitAndFilterToCTEProduceRule;
+import com.starrocks.sql.optimizer.rule.transformation.RewriteMultiDistinctRule;
+import com.starrocks.sql.optimizer.rule.transformation.SeparateProjectRule;
+import com.starrocks.sql.optimizer.rule.transformation.UnionToValuesRule;
+import com.starrocks.sql.optimizer.rule.transformation.pruner.CboTablePruneRule;
+import com.starrocks.sql.optimizer.rule.tree.PushDownAggregateRule;
+import com.starrocks.sql.optimizer.rule.tree.PushDownDistinctAggregateRule;
+import com.starrocks.sql.optimizer.task.OptimizeGroupTask;
+import com.starrocks.sql.optimizer.task.TaskContext;
+import com.starrocks.sql.optimizer.task.TaskScheduler;
+
+import java.util.List;
+
+public class SPMOptimizer extends Optimizer {
+    private final TaskScheduler scheduler = new TaskScheduler();
+    private final Memo memo = new Memo();
+    private ColumnRefSet requiredColumns;
+
+    SPMOptimizer(OptimizerContext context) {
+        super(context);
+    }
+
+    @Override
+    public OptExpression optimize(OptExpression tree, PhysicalPropertySet requiredProperty,
+                                  ColumnRefSet requiredColumns) {
+
+        context.setMemo(memo);
+        context.setTaskScheduler(scheduler);
+        this.requiredColumns = requiredColumns;
+
+        TaskContext taskContext = new TaskContext(context, requiredProperty, requiredColumns.clone(), Double.MAX_VALUE);
+        tree = optimizeByRule(tree, taskContext);
+
+        memo.init(tree);
+        memo.deriveAllGroupLogicalProperty();
+        memoOptimize(memo, taskContext);
+
+        return extractBestPlan(requiredProperty, memo.getRootGroup());
+    }
+
+    /*
+    Remove:
+    isEnableFineGrainedRangePredicate
+    isEnableRewriteGroupingsetsToUnionAll
+    isCboPushDownGroupingSet
+    isEnableStatsToOptimizeSkewJoin
+    MVRewrite
+    PruneTable
+    PRUNE_UKFK_JOIN_RULES
+     */
```

Comment:
What is the impact of removing this? Will MV rewrite be disabled if SPM is enabled? What is the purpose of SPMOptimizer?

Reply:
Because these rules based on specific parameters/predicates, use to generate the fixed-plan may not be suitable. however, this will not affect query rewrite, and after replace fixed-plan, will try to rewrite MV through the optimizer. 

we can support some hints to specify MV rewrite in the future.